File: ios/PiTools/Sources/PiTools/ReadFileTool.swift
import Foundation
import PiAgentCore
import PiAI

public struct ReadFileTool: Tool, Sendable {
    public let name = "read_file"
    public let description = "Read the contents of a file. For text files, returns the content with line numbers. For images, returns base64-encoded data. Supports optional line range."

    private let sandboxURL: URL

    public init(sandboxURL: URL) {
        self.sandboxURL = sandboxURL
    }

    public var parametersSchema: JSONValue {
        .object([
            "type": .string("object"),
            "required": .array([.string("path")]),
            "properties": .object([
                "path": .object([
                    "type": .string("string"),
                    "description": .string("Path to the file to read, relative to sandbox"),
                ]),
                "start_line": .object([
                    "type": .string("integer"),
                    "description": .string("Starting line number (1-indexed). Optional."),
                ]),
                "end_line": .object([
                    "type": .string("integer"),
                    "description": .string("Ending line number (inclusive). Optional."),
                ]),
            ]),
        ])
    }

    public func execute(input: JSONValue) async throws -> AgentToolResult {
        guard let path = input["path"]?.stringValue else {
            return AgentToolResult(output: "Error: 'path' parameter is required", isError: true)
        }

        let fileURL = sandboxURL.appendingPathComponent(path)

        // Validate path is within sandbox
        guard fileURL.standardizedFileURL.path.hasPrefix(sandboxURL.standardizedFileURL.path) else {
            return AgentToolResult(output: "Error: Path is outside sandbox", isError: true)
        }

        let fm = FileManager.default
        guard fm.fileExists(atPath: fileURL.path) else {
            return AgentToolResult(output: "Error: File not found: \(path)", isError: true)
        }

        // Check if it's an image
        let ext = fileURL.pathExtension.lowercased()
        let imageExtensions = ["png", "jpg", "jpeg", "gif", "bmp", "tiff", "webp", "heic"]
        if imageExtensions.contains(ext) {
            guard let data = fm.contents(atPath: fileURL.path) else {
                return AgentToolResult(output: "Error: Could not read file", isError: true)
            }
            let base64 = data.base64EncodedString()
            let mimeType = mimeTypeForExtension(ext)
            return AgentToolResult(
                output: "Image file: \(path) (\(data.count) bytes, \(mimeType))\nBase64: \(base64.prefix(100))...",
                details: .file(path: path, content: "[base64:\(base64)]", language: nil)
            )
        }

        // Read text file
        guard let data = fm.contents(atPath: fileURL.path),
              let content = String(data: data, encoding: .utf8)
        else {
            return AgentToolResult(output: "Error: Could not read file as text", isError: true)
        }

        let allLines = content.components(separatedBy: "\n")
        let startLine = max(1, input["start_line"]?.intValue ?? 1)
        let endLine = min(allLines.count, input["end_line"]?.intValue ?? allLines.count)

        guard startLine <= endLine else {
            return AgentToolResult(output: "Error: Invalid line range \(startLine)-\(endLine)", isError: true)
        }

        let selectedLines = allLines[(startLine - 1)..<endLine]
        var output = ""
        for (index, line) in selectedLines.enumerated() {
            let lineNum = startLine + index
            output += "\(lineNum)\t\(line)\n"
        }

        let language = languageForExtension(ext)
        let info = "File: \(path) | Lines: \(startLine)-\(endLine) of \(allLines.count) | Language: \(language ?? "unknown")"

        return AgentToolResult(
            output: "\(info)\n\(output)",
            details: .file(path: path, content: output, language: language)
        )
    }

    private func mimeTypeForExtension(_ ext: String) -> String {
        switch ext {
        case "png": return "image/png"
        case "jpg", "jpeg": return "image/jpeg"
        case "gif": return "image/gif"
        case "bmp": return "image/bmp"
        case "tiff": return "image/tiff"
        case "webp": return "image/webp"
        case "heic": return "image/heic"
        default: return "application/octet-stream"
        }
    }

    private func languageForExtension(_ ext: String) -> String? {
        switch ext {
        case "swift": return "swift"
        case "kt", "kts": return "kotlin"
        case "java": return "java"
        case "py": return "python"
        case "js": return "javascript"
        case "ts": return "typescript"
        case "rb": return "ruby"
        case "go": return "go"
        case "rs": return "rust"
        case "c", "h": return "c"
        case "cpp", "hpp", "cc": return "cpp"
        case "cs": return "csharp"
        case "json": return "json"
        case "xml": return "xml"
        case "yaml", "yml": return "yaml"
        case "toml": return "toml"
        case "md", "markdown": return "markdown"
        case "html", "htm": return "html"
        case "css": return "css"
        case "sh", "bash", "zsh": return "shell"
        case "sql": return "sql"
        case "txt": return "plaintext"
        default: return nil
        }
    }
}

File: android/pi-tools/src/main/kotlin/com/pimobile/tools/ReadFileTool.kt
package com.pimobile.tools

import android.content.ContentResolver
import android.net.Uri
import com.pimobile.agent.AgentToolResult
import com.pimobile.agent.Tool
import com.pimobile.agent.ToolResultDetails
import kotlinx.serialization.json.*
import java.io.File
import android.util.Base64

class ReadFileTool(
    private val contentResolver: ContentResolver,
    private val sandboxDir: File
) : Tool {

    override val name = "read_file"
    override val description = "Read the contents of a file. Supports sandbox paths and content:// URIs. For images, returns base64-encoded data."
    override val parametersSchema = buildJsonObject {
        put("type", "object")
        putJsonObject("properties") {
            putJsonObject("path") {
                put("type", "string")
                put("description", "File path (sandbox-relative) or content:// URI")
            }
            putJsonObject("startLine") {
                put("type", "integer")
                put("description", "Starting line number (1-based, optional)")
            }
            putJsonObject("endLine") {
                put("type", "integer")
                put("description", "Ending line number (1-based, inclusive, optional)")
            }
        }
        putJsonArray("required") { add("path") }
    }

    private val imageExtensions = setOf("jpg", "jpeg", "png", "gif", "webp", "bmp")

    private val extensionToLanguage = mapOf(
        "kt" to "kotlin", "java" to "java", "py" to "python", "js" to "javascript",
        "ts" to "typescript", "json" to "json", "xml" to "xml", "html" to "html",
        "css" to "css", "md" to "markdown", "yaml" to "yaml", "yml" to "yaml",
        "sh" to "bash", "sql" to "sql", "swift" to "swift", "rs" to "rust",
        "go" to "go", "rb" to "ruby", "c" to "c", "cpp" to "cpp", "h" to "c",
        "gradle" to "groovy", "toml" to "toml", "txt" to "text"
    )

    override suspend fun execute(input: JsonObject): AgentToolResult {
        val path = input["path"]?.jsonPrimitive?.contentOrNull
            ?: return AgentToolResult("", "Error: 'path' is required", isError = true)

        val startLine = input["startLine"]?.jsonPrimitive?.intOrNull
        val endLine = input["endLine"]?.jsonPrimitive?.intOrNull

        return try {
            if (path.startsWith("content://")) {
                readContentUri(path, startLine, endLine)
            } else {
                readSandboxFile(path, startLine, endLine)
            }
        } catch (e: Exception) {
            AgentToolResult("", "Error reading file: ${e.message}", isError = true)
        }
    }

    private fun readSandboxFile(path: String, startLine: Int?, endLine: Int?): AgentToolResult {
        val file = File(sandboxDir, path).canonicalFile
        if (!file.canonicalPath.startsWith(sandboxDir.canonicalPath)) {
            return AgentToolResult("", "Error: Path is outside sandbox", isError = true)
        }

        if (!file.exists()) {
            return AgentToolResult("", "Error: File not found: $path", isError = true)
        }

        val extension = file.extension.lowercase()

        if (extension in imageExtensions) {
            val bytes = file.readBytes()
            val base64 = Base64.encodeToString(bytes, Base64.NO_WRAP)
            val mimeType = when (extension) {
                "jpg", "jpeg" -> "image/jpeg"
                "png" -> "image/png"
                "gif" -> "image/gif"
                "webp" -> "image/webp"
                "bmp" -> "image/bmp"
                else -> "application/octet-stream"
            }
            return AgentToolResult(
                "",
                "[Image: ${file.name}, ${bytes.size} bytes, $mimeType]\nbase64:$base64"
            )
        }

        val lines = file.readLines()
        val language = extensionToLanguage[extension]

        val start = (startLine ?: 1).coerceAtLeast(1)
        val end = (endLine ?: lines.size).coerceAtMost(lines.size)

        val numberedLines = lines.subList(start - 1, end).mapIndexed { index, line ->
            "${start + index}: $line"
        }

        val content = numberedLines.joinToString("\n")
        val header = "File: $path (${lines.size} lines total, showing $start-$end)"

        return AgentToolResult(
            "",
            "$header\n$content",
            ToolResultDetails.File(path, content, language)
        )
    }

    private fun readContentUri(uriString: String, startLine: Int?, endLine: Int?): AgentToolResult {
        val uri = Uri.parse(uriString)
        val inputStream = contentResolver.openInputStream(uri)
            ?: return AgentToolResult("", "Error: Cannot open URI: $uriString", isError = true)

        val content = inputStream.bufferedReader().use { it.readText() }
        val lines = content.lines()

        val start = (startLine ?: 1).coerceAtLeast(1)
        val end = (endLine ?: lines.size).coerceAtMost(lines.size)

        val numberedLines = lines.subList(start - 1, end).mapIndexed { index, line ->
            "${start + index}: $line"
        }

        val result = numberedLines.joinToString("\n")
        return AgentToolResult(
            "",
            "URI: $uriString (${lines.size} lines total, showing $start-$end)\n$result"
        )
    }
}
